### PR TITLE
Add multiple feature with the same type but different values to product

### PR DESCRIPTION
**Pull request by fatmaBouchekoua**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | If you create a product which has the same feature twice when you reload the page the first one disappears. This PR add the possibility to create feature more than once.
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-858
| How to test?  | Reinstall Prestashop, then add the same feature to a product more than once in the BO, Go to the FO and check the product details you will see all features in the same table line.

### CODE REVIEW

**Comment by aleeks** [comment]
Hello  @fatmaBouchekoua 
Still in wip ? can you rebase please ?

**Comment by fatmaBouchekoua** [comment]
Hello @aleeks ,
i did complete working on it and i've rebased it

**Comment by aleeks** [comment]
Hello @fatmaBouchekoua ,
Thank you, let's go to test it, @vincentbz

**Comment by vincentbz** [comment]
Hi,

So it works for most of the cases, only one missing : if I have several customized values. It only displays one, not all of them (like the predefined values)

image

image

Thanks !

**Comment by aleeks** [comment]
It can not be done on 1.7.1.x, this is a improvement, not a fix. You have to do this in develop!

**Review by LittleBigDev — CHANGES_REQUESTED** [review]
```php
while (true) {
    /* things */
    if (/* test */) {
        break;
    }
}
```
should be changed to
```php
do {
    /* things */
} while (/* inversed test */);
```

File: controllers/admin/AdminProductsController.php [hunk 1]
```diff
@@ -866,19 +866,36 @@ public function processFeatures($id_product = null)
...
+                        $product->addFeaturesToDB($match[1], $value);
                     } else {
                         if ($default_value = $this->checkFeatures($languages, $match[1])) {
-                            $id_value = $product->addFeaturesToDB($match[1], 0, 1);
-                            foreach ($languages as $language) {
-                                if ($cust = Tools::getValue('custom_'.$match[1].'_'.(int)$language['id_lang'])) {
-                                    $product->addFeaturesCustomToDB($id_value, (int)$language['id_lang'], $cust);
-                                } else {
-                                    $product->addFeaturesCustomToDB($id_value, (int)$language['id_lang'], $default_value);
+                            while (true) {
```
Comment (LittleBigDev) [inline controllers/admin/AdminProductsController.php]: Please use do {} while ();

File: controllers/admin/AdminProductsController.php [hunk 2]
```diff
@@ -1576,36 +1593,44 @@ protected function checkFeatures($languages, $feature_id)
...
 
-        foreach ($languages as $language) {
-            if ($val = Tools::getValue('custom_'.$feature_id.'_'.$language['id_lang'])) {
-                $current_language = new Language($language['id_lang']);
-                if (Tools::strlen($val) > $rules['sizeLang']['value']) {
-                    $this->errors[] = $this->trans(
-                        'The name for feature %1$s is too long in %2$s.',
-                        array(
-                            ' <b>'.$feature['name'].'</b>',
+        while (true) {
```
Comment (LittleBigDev) [inline controllers/admin/AdminProductsController.php]: please use do {} while ();

**Review by aleeks — CHANGES_REQUESTED** [review]
Hello @fatmaBouchekoua , can you tell me if are you ok with this ?

File: src/Core/Product/ProductPresenter.php [hunk 3]
```diff
@@ -542,6 +542,29 @@ public function addAttributesSpecificReferences(array $product, array $presented
         return $presentedProduct;
     }
 
+    /**
+     * Assemble the same features in one array
+     * @param $presentedProduct
+     * @return mixed
+     */
+    public function addFeaturesToDisplay($presentedProduct)
```
Comment (fatmaBouchekoua) [inline src/Core/Product/ProductPresenter.php]: Done. Thanks !

File: controllers/admin/AdminProductsController.php (hunk 1 above)
Comment (fatmaBouchekoua) [inline controllers/admin/AdminProductsController.php]: Done. Thanks !

**Review by LittleBigDev — DISMISSED** [review]

File: controllers/admin/AdminProductsController.php [hunk 4]
```diff
@@ -866,20 +866,34 @@ public function processFeatures($id_product = null)
...
-                            foreach ($languages as $language) {
-                                if ($cust = Tools::getValue('custom_'.$match[1].'_'.(int)$language['id_lang'])) {
-                                    $product->addFeaturesCustomToDB($id_value, (int)$language['id_lang'], $cust);
-                                } else {
-                                    $product->addFeaturesCustomToDB($id_value, (int)$language['id_lang'], $default_value);
+                            do {
+                                $id_value = $product->addFeaturesToDB($match[1], 0, 1);
+                                $has_next_custom = false;
+                                foreach ($languages as $language) {
+                                    if ($cust = Tools::getValue('custom_' . $match[1] . '_' . (int)$language['id_lang'] . '_' . $counter)) {
```
Comment (aleeks) [inline controllers/admin/AdminProductsController.php]: Hello @fatmaBouchekoua 
Please, see your code and when you have finished your feature, try to minimize your code, 
try if this works: 

```php
foreach ($languages as $language) {
    $customFeature = Tools::getValue('custom_' . $match[1] . '_' . (int)$language['id_lang'] . '_' . $counter);
    $finalFeature = $customFeature ? $customFeature : $default_value;

    $product->addFeaturesCustomToDB($id_value, (int)$language['id_lang'], $finalFeature);

    $has_next_custom = (bool)Tools::getValue('custom_' . $match[1] . '_' . (int)$language['id_lang'] . '_' . ($counter + 1));
}
```

Be easy, 
```php 
$something = false;
//some logic
if ($somethingelse) {
    $something = true;
}

// so
$something = false
//some logic
$something = (bool) $somethingelse
```

It's really more complicated to see if you code is correct/good or not :(

File: controllers/admin/AdminProductsController.php [hunk 5]
```diff
@@ -1576,37 +1590,43 @@ protected function checkFeatures($languages, $feature_id)
...
+                    // Getting default language
+                    if ($language['id_lang'] == Configuration::get('PS_LANG_DEFAULT')) {
+                        return $value;
+                    }
                 }
-                // Getting default language
-                if ($language['id_lang'] == Configuration::get('PS_LANG_DEFAULT')) {
-                    return $val;
+                if (Tools::getValue('custom_' . $feature_id . '_' . $language['id_lang'] . '_' . ($counter + 1))) {
+                    $has_next_custom = true;
```
Comment (aleeks) [inline controllers/admin/AdminProductsController.php]: Same here 
```php 
$has_next_custom = (bool) your condition
```

File: controllers/admin/AdminProductsController.php [hunk 6]
```diff
@@ -1576,37 +1590,43 @@ protected function checkFeatures($languages, $feature_id)
...
+                    }
+                    // Getting default language
+                    if ($language['id_lang'] == Configuration::get('PS_LANG_DEFAULT')) {
+                        return $value;
+                    }
                 }
-                // Getting default language
-                if ($language['id_lang'] == Configuration::get('PS_LANG_DEFAULT')) {
-                    return $val;
+                if (Tools::getValue('custom_' . $feature_id . '_' . $language['id_lang'] . '_' . ($counter + 1))) {
```
Comment (aleeks) [inline controllers/admin/AdminProductsController.php]:
```php 
$has_next_custom = (bool)Tools::getValue('custom_' . $match[1] . '_' . (int)$language['id_lang'] . '_' . ($counter + 1))
```

**Review by aleeks — CHANGES_REQUESTED** [review]

**Review by aleeks — DISMISSED** [review]

**Comment by vincentbz** [comment]
Hi guys,

What do we need to definitively finish this PR please ? 
@aleeks  @maximebiloe  ?

**Comment by LittleBigDev** [comment]
First of all, it needs fixes so Travis and Codacy validate it.

**Comment by vincentbz** [comment]
Hi guys,

I would love to see that PR in the 1.7.3.0 version. Could we go on to finish it ? 
cadeau @eternoendless :p 

Thanks !

**Comment by fatmaBouchekoua** [comment]
Hello @vincentbz ,
I've created an other one https://github.com/PrestaShop/PrestaShop/pull/8271
So I close this one !

**Comment by vincentbz** [comment]
Ok thanks ! :)

**Pull request closed without merging**